Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1172,33 +1175,73 @@ void RenderBlockFlow::adjustPositionedBlock(RenderBox& child, const MarginInfo&
+    LayoutUnit startPosition = startPositionWithOnlyBorderAndPadding;
     if ((shouldPlaceVerticalScrollbarOnLeft() || style().scrollbarGutter().bothEdges) && isHorizontalWritingMode())
         startPosition += (style().isLeftToRightDirection() ? 1 : -1) * verticalScrollbarWidth();
     if (style().scrollbarGutter().bothEdges && !isHorizontalWritingMode())
         startPosition += (style().isLeftToRightDirection() ? 1 : -1) * horizontalScrollbarHeight();
-    LayoutUnit totalAvailableLogicalWidth = borderAndPaddingLogicalWidth() + availableLogicalWidth();
+    const LayoutUnit totalAvailableLogicalWidth = borderAndPaddingLogicalWidth() + availableLogicalWidth();
 
     LayoutUnit childMarginStart = marginStartForChild(child);
-    LayoutUnit newPosition = startPosition + childMarginStart;
+    startPosition += childMarginStart;
 
     LayoutUnit positionToAvoidFloats;
-    
     if (child.avoidsFloats() && containsFloats())
         positionToAvoidFloats = startOffsetForLine(logicalTopForChild(child), logicalHeightForChild(child));
 
+    // The non-standard WebKitCenter value should be overriden by justify-self.
+    if ((child.style().justifySelf().position() == ItemPosition::Auto && style().textAlign() == TextAlignMode::WebKitCenter) || child.style().marginStartUsing(&style()).isAuto()) {
+        // If the child is being centred then the margin calculated to do that has factored in any offset required to
+        // avoid floats, so use it if necessary.
+        startPosition = std::max(startPosition, positionToAvoidFloats + childMarginStart);
+        setLogicalLeftForChild(child, style().isLeftToRightDirection() ? startPosition : totalAvailableLogicalWidth - startPosition - logicalWidthForChild(child), applyDelta);
+        return;
+    }
+
+    auto computeSelfAlignedLogicalLeftPositionForChild = [&](LayoutUnit currentPosition) -> LayoutUnit {
+        auto startAlignedPosition = currentPosition;
+        auto endAlignedPosition = totalAvailableLogicalWidth - startAlignedPosition - logicalWidthForChild(child);
+        auto selfAlignmentStyle = child.style().resolvedJustifySelf(&style(), ItemPosition::Normal);
+        const auto parentDirection = style().direction();
+        if (!isLeftToRightDirection(parentDirection))
+            std::swap(startAlignedPosition, endAlignedPosition);
+        if (child.style().justifySelf().overflow() == OverflowAlignment::Safe && logicalWidthForChild(child) > totalAvailableLogicalWidth)
+            return startAlignedPosition;
+        const auto childDirection = child.style().direction();
+        if (selfAlignmentStyle.isStartward(childDirection, parentDirection))
+            return startAlignedPosition;
+        if (selfAlignmentStyle.isEndward(childDirection, parentDirection))
+            return endAlignedPosition;
+        if (selfAlignmentStyle.isCentered())
+            return totalAvailableLogicalWidth / 2 - logicalWidthForChild(child) / 2;
```

Comment:
```suggestion
            return (totalAvailableLogicalWidth - logicalWidthForChild(child)) / 2;
```

This avoids doing the division twice.